samples: ncp: Added NCP sample application (#2543)
nrfconnect/sdk-nrf: samples/openthread/ncp/README.rst

.. _ot_ncp_sample:

Thread: NCP sample
##################

The Thread NCP sample demonstrates the usage of OpenThread Network Co-Processor architecture inside the Zephyr shell.

Overview
********

> The sample allows to build Thread Border Router target on the MCU. 

Author: Done

> It is important to mention that, according to NCP architecture, MCU part

Reviewer: I think there is a one sentence per line rule in ncs doc: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#rst-file-formatting

applies to most of this doc

Author: Done

needs to cooperate with user higher layer process to altogether establish complete 
Thread Border Router. Recommended solution is to use docker version of OTBR (OpenThread Border Router)
and WPAN Network Daemon tool (wpantund), working under Linux operating system, for such purpose 

> (for more information read this: https://openthread.io/guides/border-router).

Reviewer: link: see comment about links below

Author: Done

There is possibility to configure number of OpenThread features, but by default
application has COMMISSIONER, JOINER, BORDER_ROUTER, BORDER_AGENT and DHCP6_SERVER
functionalities enabled.

Requirements
************

* One of the following development kits for testing the network status:

>   * |nRF52840DK|

Reviewer: I think tle look of this section was changed not long ago. Please adjust - have a look at CLI sample

Author: Done

  * |nRF52833DK|

* Border Router higher layer process communicating with MCU border router part.

* Optionally, one or more compatible development kits programmed with OpenThread samples to test communication
  in bigger network.

User interface
**************

Main recommended interaction with the application should be done via wpantund

> Linux process using its comannds (for more information: https://github.com/openthread/wpantund).

Reviewer: don't link directly. Add the link into proper section of https://github.com/nrfconnect/sdk-nrf/blob/master/doc/nrf/links.txt and ref it here.
Doc coding standard: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#hyperlinks

Author: Done

It is also possible to communicate with board using pyspinel tool via serial interface 

> (for more information: https://github.com/openthread/pyspinel). 

Reviewer: same as above

Author: Done

Building and running
********************

.. |sample path| replace:: :file:`samples/openthread/ncp`

|enable_thread_before_testing|

.. include:: /includes/build_and_run.txt

Testing
=======

To perform tests it is necessary to have at least one developement kit board
and second Thread node for checking connection. Depending on needs it might be 
another developement kit board running OpenThread stack or the simulated node on PC.

> Below steps were based on the given connectivity test: https://openthread.io/guides/border-router/docker/test-connectivity,

Reviewer: link ^

Author: Done

where detailed description may be found.

> #. First of all it is needed to build sample and programm developement kit with it.

Reviewer: Building and flashing  is covered by the Building and running section above. Steps should start after the sample is flashed unless some additional boards need to be prepared

Author: Done

#. Connect board to the PC and get its serial port name.
#. Run OTBR using docker and pass board serial port name as ncp interface.
#. Now bring up second Thread node, for example by flashing developement

>    kit board with OpenThread CLI FTD_JOINER sample (see here: https://codelabs.developers.google.com/codelabs/openthread-hardware/#4) 

Reviewer: if possible please refer to some existing ncs sample instead. Note: commands below will need to be adjusted

.. _ot_ncp_sample:

Thread: NCP
###########

> The Thread NCP sample demonstrates the usage of OpenThread Network Co-Processor architecture inside the Zephyr shell.

Reviewer 2: I think we should use something like "Zephyr environment", Zephyr's shell is not enabled here

Author: Done

The sample is based on Zephyr's :ref:`zephyr:ncp-sample` sample.
However, it customizes Zephyr's sample to the NCS requirements (for example, by increasing the stack size dedicated for the user application), and also extends it with several features:

* Increased mbedTLS heap size.

> * Lowered main stack size to increase user application space

Reviewer 2: nit: dot

Author: Done

The sample shows how to set connection between NCP and wpantund.

This sample has Commissioner, Joiner, and Border Router roles enabled by default.
Additionally, it comes with the following functionalities enabled:

* :option:`CONFIG_OPENTHREAD_BORDER_AGENT` - Enables Border Agent support.
* :option:`CONFIG_OPENTHREAD_DHCP6_SERVER` - Enables DHCPv6 server capability.

Requirements
************

The sample supports the following development kits for testing the network status:

.. include:: /includes/boardname_tables/sample_boardnames.txt
   :start-after: set8_start
   :end-before: set8_end

To test the sample, you need at least one development kit NCP board.
Additional NCP boards can be used for the :ref:`optional testing of network joining <ot_ncp_sample_testing_more_boards>`.

Moreover, the sample requires a Userspace higher layer process running on user's device in order to communicate with the MCU NCP part.
This sample uses `wpantund`_ as reference.

User interface
**************

All the interactions with the application are handled using serial communication.

> For the interaction with the application, this sample uses `wpantund`_ process with its commands.

Reviewer 2: You control wpantund with dbus, 'wpanctl' is the tool which provides command interface.

Author: Done

> It is also possible to communicate with NCP board using `pyspinel`_ commands through serial interface.

Reviewer 2: I think that "through serial interface" is redundant here

Author: Done

You can use your own application instead of wpantund and pyspinel provided that it supports the spinel communication protocol.

> Application by default uses baudrate 1000000 bit/s for serial communication.

Reviewer 2: I think that here we should refer rather to "example" than "Application" and rather "reconfigures" than "uses"

Author: Done

Building and running
********************

.. |sample path| replace:: :file:`samples/openthread/ncp`

|enable_thread_before_testing|

.. include:: /includes/build_and_run.txt

Testing
=======

After building the sample and programming it to your development kit, test it by performing the following steps:

1. Connect the NCP board's SEGGER J-Link USB port to the PC USB port with an USB cable.

> #. Get the board's serial port name.

Reviewer 2: I think it's worth to add "(e.g. /dev/ttyACM0)"

Author: Done

.. _ot_ncp_sample:

> Thread: NCP sample

Reviewer 3: ```suggestion
Thread: NCP
```
`sample` should never be part of the sample name.

Author: Done

##################

> The Thread NCP sample demonstrates the usage of OpenThread Network Co-Processor architecture inside the Zephyr shell.

Author: Done

Overview
********

The sample demonstrates usage of NCP target on the MCU to communicate it with user WPAN Network Daemon, allowing to create for example Thread Border Router.
It is important to mention that, according to NCP architecture, MCU part needs to cooperate with user higher layer process to altogether establish complete Thread Border Router.

> Recommended solution is to use docker version of OTBR (OpenThread Border Router) and WPAN Network Daemon tool (wpantund), working under Linux operating system, for such purpose (for more information go there: `OpenThread Border Router`_).

Reviewer 3: Does wpantund work off-the-shelf with NCS? No special configuration needed for NCS?

Author: It's not the matter of NCS, but NCP architecture by design works with wpantund via spinel protocol - I guess. @MarekPorwisz is this right?

Reviewer 3: I'm asking also because wpantund needs to be added to the list of Thread tools in the Thread user guide.

> Recommended solution is to use docker version of OTBR (OpenThread Border Router) and WPAN Network Daemon tool (wpantund), working under Linux operating system, for such purpose (for more information go there: `OpenThread Border Router`_).

Reviewer 3: What is the OTBR docker? Please explain or provide a link to the explanation.

Author: I rephrased it, because OTBR is just exemplary solution and I should have been explain this. Anyway I'm not sure what do you mean. Should I clarify what is docker? I wrote what is OTBR in brackets and the link for explanation with install tutorial is also attached.

> Recommended solution is to use docker version of OTBR (OpenThread Border Router) and WPAN Network Daemon tool (wpantund), working under Linux operating system, for such purpose (for more information go there: `OpenThread Border Router`_).

Author: Done

> There is possibility to configure number of OpenThread features, but by default application has COMMISSIONER, JOINER, BORDER_ROUTER, BORDER_AGENT and DHCP6_SERVER functionalities enabled.

Reviewer 3: This needs to be rephrased.
Supported OT features are mentioned on this page: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_thread.html
If this sample does not support any of these features, we need to mention these unsupported features in a note.
If this sample supports features that are not on the list on the page I linked, then does this mean this sample introduces support for new features and they need to be added to that list?

Author: I think it does support all Thread features and does not introduce new one (at least for now). I just was basing on the NCS CLI sample, where also such things were written in the .rst file. If it is wrong, maybe it should be changed in both samples?

Requirements
************

The sample supports the following development kits for testing the network status:

.. include:: /includes/boardname_tables/sample_boardnames.txt
   :start-after: set8_start
   :end-before: set8_end

> Border Router higher layer process is needed to communicate with MCU border router part.

Reviewer 3: What process are we talking here? Is it something you need to enable? If so, how and where? If it is enabled by default, can you mention this here?

Author: Done

Optionally, one or more compatible development kits programmed with OpenThread samples to test communication in bigger network.

User interface
**************

> Main recommended interaction with the application should be done via wpantund Linux process using its comannds (for more information check: `wpantund usage`_).

Reviewer 3: Typo: `comannds` > `commands`

Author: Done

> Main recommended interaction with the application should be done via wpantund Linux process using its comannds (for more information check: `wpantund usage`_).

Reviewer 3: "Recommended" can suggest some interactions are not recommended or discouraged. If so, we need to be more specific here.

Author: Done

> It is also possible to communicate with board using pyspinel tool via serial interface (for more information check: `pyspinel commands`_).

Reviewer 3: You mean the NCP board or any board, including the optional ones?

Author: Only NCP. I added it to clarify.

Building and running
********************

.. |sample path| replace:: :file:`samples/openthread/ncp`

|enable_thread_before_testing|

.. include:: /includes/build_and_run.txt

Testing
=======

To perform tests it is necessary to have at least one development kit board and second Thread node for checking connection.

> Depending on needs it might be another development kit board running OpenThread stack or the simulated node on PC.

Author: I moved this to requirements and clarified a bit.

> Below steps were based on the given connectivity test: `OpenThread Border Router connectivity test`_, where detailed description may be found.

Author: Outdated. There are not any links to outer tutorials anymore.

> #. Connect board to the PC and get its serial port name.

Reviewer 3: Connect how exactly? Where can I get the serial port name from?

Author: First part done. Second part - I don't understand. Should we explain how to check connected device COM port in the system?

> #. Run OTBR using docker and pass board serial port name as ncp interface.

Reviewer 4: s/ncp/NCP

> #. Now bring up second Thread node, for example by flashing development kit board with OpenThread CLI FTD_JOINER sample (see here: `Codelabs - Build Thread network`_)

Reviewer 4: Can't we recommend to use NCS CLI sample instead? Should be way simpler in this case than building a sample from OT repo.

Author: Done

> #. Now bring up second Thread node, for example by flashing development kit board with OpenThread CLI FTD_JOINER sample (see here: `Codelabs - Build Thread network`_)

Reviewer 3: Can you provide example commands for each of these steps?

Author: Done. I provided commands with exemplary outputs in every step.

> #. Now bring up second Thread node, for example by flashing development kit board with OpenThread CLI FTD_JOINER sample (see here: `Codelabs - Build Thread network`_)

Author: I think it's fixed.

#. Invoke some of the OpenThread commands to make new node to join to the border router network:

  .. code-block:: console

    uart:~$ dataset panid <network_panid>
    uart:~$ dataset masterkey <network_masterkey>
    uart:~$ dataset commit active
    uart:~$ ifconfig up
    uart:~$ thread start

>     Parameters: network_panid and network_masterkey can be read from the OTBR process output logs

Reviewer 3: Where I can find these logs?

>     Parameters: network_panid and network_masterkey can be read from the OTBR process output logs

Author: Done

> #. At this point, typing "state" command into the node CLI interface should end with answer "child" or "router" depending on running time.

Reviewer 3: Please rephrase this step into a command: "Type ``state``..."

Author: Done

> #. Invoke some external IPv6 address 16 bytes from 64:ff9b:0:0:0:0:acd9:a46e: icmp_seq=1 hlim=118 time=45mss ping command to prove that it is possible to communicate with hosts from outside of our network:

Reviewer 3: ```suggestion
#. Run the following command to verify that it is possible to communicate with hosts from outside of the network:
```

Author: Done

  .. code-block:: console

    uart:~$ ping 64:ff9b::acd9:a46e

>     16 bytes from 64:ff9b:0:0:0:0:acd9:a46e: icmp_seq=1 hlim=118 time=45ms

Reviewer 3: I understand that the second line is the result of the ping?
If so, we need to introduce the result:
```suggestion
This command pings the device at the provided address.
If the ping is successful, you will receive the following output (or similar):

.. code-block:: console

   16 bytes from 64:ff9b:0:0:0:0:acd9:a46e: icmp_seq=1 hlim=118 time=45ms

```

Author: Done

Dependencies
************

This sample uses the following Zephyr libraries:

* :ref:`zephyr:kernel_api`:

  * ``include/kernel.h``

* :ref:`zephyr:thread_protocol_interface`

* :ref:`zephyr:logging_api`:

>   * ``include/logging/log.h``

Reviewer 3: ```suggestion
  * ``include/logging/log.h``

----

Copyright disclaimer
    |Google_CCLicense|

```
You need an extra blank line at the end, and the copyright disclaimer, as you're referring to the OpenThread user guide.

Author: Done
